lib/pavilion/schedulers/plugins/flux.py

```diff
@@ -0,0 +1,291 @@
+from ..config import validate_list
+
+
+# Just import flux once
+try:
+    import flux
+    import flux.job
+    import flux.resource
+    from flux.job import JobspecV1
+except ImportError:
+    flux = None
+
+flux_states = [
+    "DEPEND",
+    "SCHED",
+    "RUN",
+    "CLEANUP",
+    "COMPLETED",
+    "FAILED",
+    "CANCELED",
+    "TIMEOUT",
+]
+
+
+class Flux(SchedulerPluginAdvanced):
+    """
+    Schedule tests with Flux!
+    """
+
+    def __init__(self):
+        super().__init__("flux", "Schedules tests via the Flux Framework scheduler.")
+
+    def _get_config_elems(self):
+        elems = [
+            yc.ListElem(
+                name="avail_states",
+                sub_elem=yc.StrElem(),
+                help_text="When looking for immediately available "
+                "nodes, they must be in one of these "
+                "states.",
+            ),
+            yc.ListElem(
+                name="up_states",
+                sub_elem=yc.StrElem(),
+                help_text="When looking for nodes that could be  "
+                "allocated, they must be in one of these "
+                "states.",
+            ),
+        ]
+
+        defaults = {
+            "up_states": flux_states,
+            "avail_states": flux_states,
+        }
+
+        validators = {
+            "up_states": validate_list,
+            "avail_states": validate_list,
+        }
+
+        return elems, validators, defaults
+
+    @classmethod
+    def parse_node_list(cls, node_list) -> NodeList:
+        """
+        Convert a list of strings into a list of nodes
+        """
+        return NodeList(node_list)
+
+    def _get_alloc_nodes(self, job) -> NodeList:
+        """
+        Get the list of allocated nodes.
+        """
+        listing = flux.job.list.JobList(flux.Flux(), ids=[job.info["jobid"]])
+        jobs = listing.jobs()
+        if not jobs:
+            return []
+        nodes = jobs[0]._nodelist
+        if not isinstance(nodes, list):
+            nodes = [nodes]
+        return nodes
+
+    def _get_raw_node_data(self, sched_config) -> Tuple[Union[List[Any], None], Any]:
+        """
+        Get a flux resource list
+        """
+
+        rpc = flux.resource.list.resource_list(flux.Flux())
+        listing = rpc.get()
+
+        nodelist = [str(node) for node in listing.free.nodelist]
+        # Flux doesn't currently support reservations
+        extra = {"reservations": {}, "nodes_listing": listing}
+        return nodelist, extra
+
+    def _transform_raw_node_data(self, sched_config, node_data, extra) -> NodeInfo:
+        """
+        Translate the gathered data into a NodeInfo dict.
+        """
+        listing = extra["nodes_listing"]
+        del extra["nodes_listing"]
+        node_info = NodeInfo({})
+        node_info["name"] = node_data
+        node_info["cpus"] = listing.free.ncores
+        node_info["states"] = None
+        node_info["up"] = node_data
+        node_info["available"] = node_data
```

Comment:
These have a very important and specific meaning (from the parent method doc string):

>        # Node Status - (required)
>
>        - up (bool) - Whether the node is up (allocatable).
>        - available (bool) - Whether the node is allocatable and unallocated.

Reply:
So the sets overlap?

Reply:
okay we actually have both free and up lists to match nicely to these two!